src/cost_estimates/fee_scalar.rs

```diff
@@ -0,0 +1,207 @@
+                warn!("Error in fee estimator fetching current estimates"; "err" => ?e);
+                return;
+            }
+        };
+
+        debug!("Updating fee rate estimate for new block";
+               "new_measure_fast" => new_measure.fast,
+               "new_measure_medium" => new_measure.medium,
+               "new_measure_slow" => new_measure.slow,
+               "new_estimate_fast" => next_estimate.fast,
+               "new_estimate_medium" => next_estimate.medium,
+               "new_estimate_slow" => next_estimate.slow);
+
+        let sql = "INSERT OR REPLACE INTO scalar_fee_estimator
+                     (estimate_key, fast, medium, slow) VALUES (?, ?, ?, ?)";
+
+        let tx = tx_begin_immediate_sqlite(&mut self.db).expect("SQLite failure");
+
+        tx.execute(
+            sql,
+            rusqlite::params![
+                SINGLETON_ROW_ID,
+                u64_to_sql(next_estimate.fast).unwrap_or(i64::MAX),
+                u64_to_sql(next_estimate.medium).unwrap_or(i64::MAX),
+                u64_to_sql(next_estimate.slow).unwrap_or(i64::MAX)
+            ],
+        )
+        .expect("SQLite failure");
+
+        tx.commit().expect("SQLite failure");
+    }
+}
+
+impl<M: CostMetric> FeeEstimator for ScalarFeeRateEstimator<M> {
+    fn notify_block(&mut self, receipt: &StacksEpochReceipt) -> Result<(), EstimatorError> {
+        let mut all_fee_rates: Vec<_> = receipt
+            .tx_receipts
+            .iter()
+            .filter_map(|tx_receipt| {
+                let (payload, fee, tx_size) = match tx_receipt.transaction {
+                    TransactionOrigin::Stacks(ref tx) => {
+                        Some((&tx.payload, tx.get_tx_fee(), tx.tx_len()))
+                    }
+                    TransactionOrigin::Burn(_) => None,
+                }?;
+                let scalar_cost = match payload {
+                    TransactionPayload::TokenTransfer(_, _, _) => {
+                        // TokenTransfers *only* contribute tx_len, and just have an empty ExecutionCost.
+                        self.metric.from_len(tx_size)
+                    }
+                    TransactionPayload::Coinbase(_) => {
+                        // Coinbase txs are "free", so they don't factor into the fee market.
+                        return None;
+                    }
+                    TransactionPayload::PoisonMicroblock(_, _)
+                    | TransactionPayload::ContractCall(_)
+                    | TransactionPayload::SmartContract(_) => {
+                        // These transaction payload types all "work" the same: they have associated ExecutionCosts
+                        // and contibute to the block length limit with their tx_len
+                        self.metric
+                            .from_cost_and_len(&tx_receipt.execution_cost, tx_size)
+                    }
+                };
+                let fee_rate = cmp::max(1, fee / cmp::max(1, scalar_cost));
```

Comment:
Just want to confirm that `scalar_cost` can never be zero here, right?

Reply:
Also, because fees can be low (i.e. in small numbers of uSTX), should `fee_rate` be a float here?  We're not worried about any loss of precision that might arise?

Reply:
Even if it could, this wouldn't cause an overflow, because of the `cmp::max`'s.

Reply:
Fee rate is microstacks per (1/60k)th of the block limit. So a typical contract-call would be about 300 of the execution cost units, meaning that an integer fee rate would have a precision of 300 microstacks (0.0003 stacks). At worst (e.g., a block-filling tx), the precision would be 0.06 stacks. A float could provide more precision, though I don't think it would add much. However, it's easy enough to change to a float, so I'll just do that.

Reply:
Fair enough